File: 3682-count-the-number-of-arrays-with-k-matching-adjacent-elements/count-the-number-of-arrays-with-k-matching-adjacent-elements.java
class Solution {
    static final int MOD = 1_000_000_007;
    static long[] factorial = new long[100001]; // can be increased based on constraint
    static long[] inverseFactorial = new long[100001];
    static boolean precomputed = false;

    public int countGoodArrays(int n, int m, int k) {
        if (!precomputed) {
            precomputeFactorials(100000); // safe limit for constraints
        }

        long res = nCr(n - 1, k);
        res = (res * m) % MOD;
        res = (res * modPow(m - 1, n - k - 1)) % MOD;
        return (int) res;
    }

    private void precomputeFactorials(int limit) {
        factorial[0] = 1;
        for (int i = 1; i <= limit; i++) {
            factorial[i] = (factorial[i - 1] * i) % MOD;
        }

        inverseFactorial[limit] = modPow(factorial[limit], MOD - 2); // Fermat's Little Theorem
        for (int i = limit - 1; i >= 0; i--) {
            inverseFactorial[i] = (inverseFactorial[i + 1] * (i + 1)) % MOD;
        }

        precomputed = true;
    }

    private long nCr(int n, int r) {
        if (r < 0 || r > n) return 0;
        return factorial[n] * inverseFactorial[r] % MOD * inverseFactorial[n - r] % MOD;
    }

    private long modPow(long base, long exp) {
        long result = 1;
        base %= MOD;
        while (exp > 0) {
            if ((exp & 1) == 1) {
                result = (result * base) % MOD;
            }
            base = (base * base) % MOD;
            exp >>= 1;
        }
        return result;
    }
}

File: 3682-count-the-number-of-arrays-with-k-matching-adjacent-elements/count-the-number-of-arrays-with-k-matching-adjacent-elements.c
#define MOD 1000000007
#define MAX_N 100005

long long factorial[MAX_N];
long long fermatFact[MAX_N];

// Binary exponentiation (iterative)
long long findPower(long long a, long long b) {
    long long result = 1;
    a %= MOD;
    while (b > 0) {
        if (b & 1) result = (result * a) % MOD;
        a = (a * a) % MOD;
        b >>= 1;
    }
    return result;
}

// nCr using Fermat's little theorem
long long nCr(int n, int r) {
    return factorial[n] * fermatFact[n - r] % MOD * fermatFact[r] % MOD;
}

int countGoodArrays(int n, int m, int k) {
    // Precompute factorials
    factorial[0] = factorial[1] = 1;
    for (int i = 2; i <= n; ++i) {
        factorial[i] = (factorial[i - 1] * i) % MOD;
    }

    // Precompute modular inverses using Fermat’s little theorem
    for (int i = 0; i <= n; ++i) {
        fermatFact[i] = findPower(factorial[i], MOD - 2);
    }

    long long result = nCr(n - 1, k);
    result = result * m % MOD;
    result = result * findPower(m - 1, n - k - 1) % MOD;

    return (int)result;
}
